### DISCUSSION TITLE
Declcd - A Declarative Continuous Delivery Toolkit For Kubernetes (GitOps)

### DISCUSSION THREAD

**Discussion by kharf** [cue-lang/cue#3163] [Show and tell]
Hey all,

When I first stumbled upon CUE, I knew I had to give it a shot.
For many years, I've been working with Kubernetes and IaC, but I've always found YAML frustrating to work with.
I'm a long-time fan of FluxCD and always wondered what it needs to run a GitOps tool - so, starting my own from scratch with CUE seemed like a great idea.

The result: **Declcd** (https://github.com/kharf/declcd) 

For over a year I have been working on it in my free time and I wanted to show it here, because I am very excited with the direction CUE is going and I feel like I want to make Declcd a fully fledged GitOps-Toolkit, where YAML will be kept far far away.

Thank you for this lovely language!!

**Comment by infogulch** [comment]
Hey declcd looks pretty neat! It seems to be built as glue between simple parts (go, cue, git, helm) and I like that approach.

A small nit, I noticed that the indentation on `infrastructure/prometheus.cue` was inconsistent. https://github.com/kharf/declcd?tab=readme-ov-file#deploy-a-manifest-and-a-helmrelease

[timoni](https://github.com/stefanprodan/timoni) is also intended to deploy k8s packages defined in CUE. If you've seen it, how would you compare timoni and declcd?

**Reply by kharf** [reply]
Thank you! Yea, it adheres to the GitOps principles: https://opengitops.dev/.

Documentation is very rudimentary and still very wip, but the indentation looks off, true!

Timoni is great, I have yet to integrate it. In case of Declcd, Timoni acts as a Kubernetes package manager like Helm. You would create Timoni modules and bundle Kubernetes manifests together, like Helm Charts, and define the corresponding wanted Timoni instance state in Declcd. Declcd would pick this up and apply the wanted state to the current state (install).

**Reply by infogulch** [reply]
Oh I see. So declcd is more like FluxCD based on CUE, is that right?

**Reply by kharf** [reply]
yes exactly!

**Comment by DavidGamba** [comment]
How does it differ from applying things with Carvel's Kapp? Have you ever used that workflow? Render the CUE into yaml then use Kapp to apply your changes and show you a diff.

**Reply by kharf** [reply]
Never worked with it, but from a first glance and from a high level perspective: Declcd integrates CUE natively. You dont have to worry about yaml at all. Additionally Declcd is purely declarative (GitOps by heart), whereas Kapp looks more imperative (just a guess by looking at it).
Declcd pulls the wanted state from Git, builds K8s unstructured objects from CUE definitions and applies them on the cluster automatically. It does that by decrypting secrets, detecting drifts through SSA and collecting garbage (uninstalling undeclared resources). So its more comparable to Fluxcd or Argocd, where you just commit your declarations and the rest is done for you by the controller.

**Reply by DavidGamba** [reply]
Thanks for taking the time to clarify!

**Reply by kharf** [reply]
Always!

**Comment by the-nurk** [comment]
hello we'd like to chat! kharf@nirk.io
